### PR TITLE
SIP-135 Refactor Collateral contracts for L2 shorts

### PR DESCRIPTION
**Pull request by barrasso**
Tracking tasks here: https://github.com/Synthetixio/issues/issues/264

### CODE REVIEW

File: publish/ovm-ignore.json [hunk 1]
```diff
@@ -4,16 +4,14 @@
...
-	"CollateralErc20",
-	"CollateralEth",
-	"CollateralShort",
 	"ExchangerWithVirtualSynth",
 	"FakeTradingRewards",
 	"MockBinaryOptionMarket",
 	"MockBinaryOptionMarketManager",
 	"TestableBinaryOptionMarket",
 	"Synthetix",
+	"CollateralEth",
```
Comment (justinjmoses) [inline publish/ovm-ignore.json:13]: please sort a-z

Comment (barrasso) [reply publish/ovm-ignore.json:13]: Resolved

**Review by jacko125 — DISMISSED** [review]
looks good the changes , just some comments to look at

File: contracts/Collateral.sol [hunk 2]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         _checkLoanAvailable(loan);
 
-        // 4. Check that the repayer is the borrower.
-        require(loan.account == repayer, "Must be borrower");
-
-        // 5. Accrue interest.
-        loan = accrueInterest(loan);
-
-        // 6. Repay the accruedInterest if payInterest == true.
+        // 2. Repay the accruedInterest if payInterest == true.
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol:520]: Yes: https://github.com/Synthetixio/synthetix/blob/91de78d1060b81f43cfdaa05af702bd69f3ce6f2/test/contracts/CollateralShort.js#L364

File: contracts/Collateral.sol [hunk 3]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         (uint expectedAmount, uint fee, ) = _exchanger().getAmountsForExchange(payment, loan.currency, sUSD);
 
-        // 8. Reduce the collateral by the amount repaid (minus the exchange fees).
+        // 5. Reduce the collateral by the amount repaid (minus the exchange fees).
         loan.collateral = loan.collateral.sub(expectedAmount);
 
-        // 9. Process the payment and pay the exchange fees if needed.
-        loan = _processPayment(loan, payment);
+        // 6. Process the payment and pay the exchange fees if needed.
+        _processPayment(loan, payment);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol:535]: Sounds like a good idea to close out the loan if the amount and interest is fully paid off. Should we also consider adding the same functionality to the vanilla `repay` function?

File: publish/releases.json [hunk 4]
```diff
@@ -293,6 +293,22 @@
 			"layer": "base",
 			"released": "base"
 		},
+		{
+			"sip": 135,
+			"layer": "both",
+			"sources": {
+				"base": ["SystemSettings"],
```
Comment (barrasso) [inline publish/releases.json:300]: @justinjmoses 

The L1 integration fork test fails [here](https://app.circleci.com/pipelines/github/Synthetixio/synthetix/8537/workflows/6baa513c-f221-4483-84c9-54f24b90df6f/jobs/83912?invite=true#step-106-2196) due to changes to the `Collateral` contracts (i.e. because we added the `loans` mapping). 

Should we add the relevant `Collateral` contracts to the `"base"` sources list in `releases.json`, or is there another option like using conditionals in the deploy script?

Note: including the `SystemSettings` contract in `"base"` isn't an issue since the changes to it are just additions, but the `Collateral` contracts are a different case because of the refactor.

File: contracts/Collateral.sol [hunk 5]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-        _systemStatus().requireIssuanceActive();
-
-        // 1. Check the payment amount.
-        require(payment > 0, "Payment must be above 0");
-
-        // 2. Get loan.
-        Loan memory loan = state.getLoan(borrower, id);
+    ) internal rateIsValid returns (uint, uint) {
+        // 0. Get the loan to repay and accrue interest.
+        Loan storage loan = _getLoanAndAccrueInterest(id, repayer);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 6]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
-        // 1. Check the payment amount.
-        require(payment > 0, "Payment must be above 0");
-
-        // 2. Get the loan.
-        Loan memory loan = state.getLoan(borrower, id);
-
-        // 3. Check loan is open and last interaction time.
+        // 1. Check loan is open and last interaction time.
         _checkLoanAvailable(loan);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 7]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
-        // 1. Get the loan.
-        Loan memory loan = state.getLoan(msg.sender, id);
+    function _withdrawInternal(uint id, uint amount) internal rateIsValid returns (uint, uint) {
+        // 0. Get the loan and accrue interest.
+        Loan storage loan = _getLoanAndAccrueInterest(id, msg.sender);
 
-        // 2. Check loan is open and last interaction time.
+        // 1. Check loan is open and last interaction time.
         _checkLoanAvailable(loan);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 8]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         uint id,
         uint payment
     ) internal rateIsValid returns (uint collateralLiquidated) {
-        // 0. Check the system is active.
-        _systemStatus().requireIssuanceActive();
+        // 0. Get the loan and accrue interest.
+        Loan storage loan = _getLoanAndAccrueInterest(id, borrower);
 
-        // 1. Check the payment amount.
-        require(payment > 0, "Payment must be above 0");
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol:505]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol:505]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 9]
```diff
@@ -687,77 +576,55 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+            uint interestPaid = payment > loan.accruedInterest ? loan.accruedInterest : payment;
+            loan.accruedInterest = loan.accruedInterest.sub(interestPaid);
             payment = payment.sub(interestPaid);
 
-            _payFees(interestPaid, loanBefore.currency);
+            _payFees(interestPaid, loan.currency);
         }
 
-        // If there is more payment left after the interest, pay down the principal.
+        // If there is more payment left after the interest, pay down the pr incipal.
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 10]
```diff
@@ -413,268 +357,213 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         require(amount > 0, "Deposit must be above 0");
 
-        // 2. Get the loan
-        Loan memory loan = state.getLoan(account, id);
+        // 2. Get the loan.
+        // Owner is not important here, as it is a donation to the collateral of the loan
+        Loan storage loan = loans[id];
 
         // 3. Check loan is open and last interaction time.
         _checkLoanAvailable(loan);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/CollateralShort.sol [hunk 11]
```diff
@@ -41,47 +36,47 @@ contract CollateralShort is Collateral {
...
 
     function repayWithCollateral(
         address borrower,
         uint id,
         uint amount,
         bool payInterest
-    ) external {
-        repayWithCollateralInternal(borrower, msg.sender, id, amount, payInterest);
+    ) external returns (uint short, uint collateral) {
+        (short, collateral) = _repayWithCollateralInternal(borrower, msg.sender, id, amount, payInterest);
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/CollateralShort.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/dbeff854a52f2395cbb095a58d7c7abb88b34bc5

File: contracts/Collateral.sol [hunk 12]
```diff
@@ -65,8 +65,7 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
     uint public maxLoansPerAccount = 50;
 
     // Time in seconds that a user must wait between interacting with a loan.
-    // Provides front running and flash loan protection.
-    uint public interactionDelay = 300;
+    uint public interactionDelay = 0;
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/ce31cf7ffd607cc42576fbb14c582975b483d9bb

[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: There are two major reasons for this:

1 - We don't expect frontrunning interactions to be that much of issue on L2 right now (we can always change this value in the future via `SystemSettings` if it does become a problem)

2 - Since the transactions on L2 are near-instant, adding a significant delay will negatively impact the trading UX.

[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/ce31cf7ffd607cc42576fbb14c582975b483d9bb

File: test/contracts/CollateralEth.js [hunk 13]
```diff
@@ -37,26 +30,25 @@ contract('CollateralEth', async accounts => {
...
 	const oneHundredsUSD = toUnit(100);
-	const fiveHundredSUSD = toUnit(500);
+	const fiveHundredsUSD = toUnit(500);
 
 	let tx;
 	let loan;
 	let id;
 
-	const [deployerAccount, owner, oracle, , account1, account2] = accounts;
+	const [, owner, oracle, , account1, account2] = accounts;
```
[the remark this answers is not in the document]
Comment (barrasso) [reply test/contracts/CollateralEth.js]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/60a68138ac8ab0e10ab84734027af9212c73f88e

File: contracts/SystemSettings.sol [hunk 14]
```diff
@@ -305,6 +325,53 @@ contract SystemSettings is Owned, MixinSystemSettings, ISystemSettings {
         emit EtherWrapperBurnFeeRateUpdated(_rate);
     }
 
+    function setMinCratio(address _collateral, uint _minCratio) external onlyOwner {
+        require(_minCratio > SafeDecimalMath.unit(), "Cratio must be above 1");
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/SystemSettings.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/60a68138ac8ab0e10ab84734027af9212c73f88e

File: contracts/Collateral.sol [hunk 15]
```diff
@@ -221,38 +224,9 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-
-    function setCanOpenLoans(bool _canOpenLoans) external onlyOwner {
-        canOpenLoans = _canOpenLoans;
-        emit CanOpenLoansUpdated(canOpenLoans);
-    }
-
     /* ---------- LOAN INTERACTIONS ---------- */
 
-    function openInternal(
+    function _openInternal(
```
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/2690b533a2162fcc71ceb90ef7517c279664722b

File: publish/releases.json (hunk 4 above)
Comment (barrasso) [reply publish/releases.json:300]: Resolved by adding [utils/loans.js](https://github.com/Synthetixio/synthetix/blob/2db88d4ea5f742bd65bcda8ccd6fb6d8da2a9e7c/test/integration/utils/loans.js) which contains a helper function called `getLoan` that is able to differentiate between getting loans on mainnet vs ovm.

Keeping it simple in `short.behavior.js`: 

```javascript
loan = await getLoan({ ctx, id: loanId, user });
```

File: test/integration/utils/loans.js [hunk 16]
```diff
@@ -0,0 +1,16 @@
+async function getLoan({ ctx, id, user }) {
+	const { CollateralShort, CollateralStateShort } = ctx.contracts;
+
+	let loan;
+
+	if (!ctx.useFork) {
```
Comment (justinjmoses) [inline test/integration/utils/loans.js:6]: as discussed offline, i think it's better to check see if the contract you expect exists and has the ABI function you desire (aiming for the latest version first) and if not, use the legacy approach (rather than coding in the logic that forking mainnet is legacy and otherwise it's latest)

File: contracts/Collateral.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacko125) [reply contracts/Collateral.sol:531]: I think that would be worth adding the check in to `repay` as well. Otherwise the position is technically still not closed.

File: contracts/Collateral.sol [hunk 17]
```diff
@@ -58,15 +58,8 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-    // The fee charged for issuing a loan.
-    uint public issueFeeRate;
-
-    // The maximum number of loans that an account can create with this collateral.
-    uint public maxLoansPerAccount = 50;
-
     // Time in seconds that a user must wait between interacting with a loan.
-    // Provides front running and flash loan protection.
-    uint public interactionDelay = 300;
+    uint public interactionDelay;
```
Comment (justinjmoses) [inline contracts/Collateral.sol:62]: why is this here and in `SystemSettings`?

File: contracts/Collateral.sol [hunk 18]
```diff
@@ -189,12 +186,16 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
     }
 
     // We set the interest index to 0 to indicate the loan has been closed.
-    function _checkLoanAvailable(Loan memory _loan) internal view {
-        require(_loan.interestIndex > 0, "Loan does not exist");
-        require(_loan.lastInteraction.add(interactionDelay) <= block.timestamp, "Recently interacted");
+    function _checkLoanAvailable(Loan memory loan) internal view {
+        _isLoanOpen(loan.interestIndex);
+        require(loan.lastInteraction.add(interactionDelay) <= block.timestamp, "Recently interacted");
```
Comment (justinjmoses) [inline contracts/Collateral.sol:191]: I think you mean `getInteractionDelay()` no?

File: contracts/Collateral.sol [hunk 19]
```diff
@@ -328,78 +295,53 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+        // 1. Store this for the event.
+        amount = loan.amount;
 
-        // 3. Return collateral to the child class so it knows how much to transfer.
+        // 2. Return collateral to the child class so it knows how much to transfer.
         collateral = loan.collateral;
 
-        // 4. Burn the synths
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(liquidator, loan.currency), "Waiting or owing");
+        // 3. Check that the sender has enough synths.
```
Comment (justinjmoses) [inline contracts/Collateral.sol:341]: should this not be `liquidator` not `msg.sender`?

File: contracts/Collateral.sol [hunk 20]
```diff
@@ -413,268 +355,211 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
     }
 
-    function liquidateInternal(
+    function _liquidate(
         address borrower,
         uint id,
         uint payment
     ) internal rateIsValid returns (uint collateralLiquidated) {
-        // 0. Check the system is active.
-        _systemStatus().requireIssuanceActive();
```
Comment (justinjmoses) [inline contracts/Collateral.sol:502]: [np] moving this check into `_getLoanAndAccrueInterest` is a bit counter-intuitive. I'd prefer a private funtion and modifier to state that this `requiresIssuanceActive`

File: contracts/Collateral.sol [hunk 21]
```diff
@@ -413,268 +355,211 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+
+        // 10. Work out how much collateral to redeem.
         collateralLiquidated = _collateralUtil().collateralRedeemed(loan.currency, amountToLiquidate, collateralKey);
         loan.collateral = loan.collateral.sub(collateralLiquidated);
 
-        // 13. Update the last interaction time.
-        loan.lastInteraction = block.timestamp;
-
-        // 14. Burn the synths from the liquidator.
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(msg.sender, loan.currency), "Waiting or owing");
```
Comment (justinjmoses) [inline contracts/Collateral.sol:548]: Why don't we need this check anymore?

File: contracts/Collateral.sol [hunk 22]
```diff
@@ -413,268 +355,211 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
-        // 7. Update the last interaction time.
+        // 5. Process the payment.
+        _processPayment(loan, payment);
+
+        // 6. Update the last interaction time.
         loan.lastInteraction = block.timestamp;
 
-        // 8. Burn synths from the payer
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(repayer, loan.currency), "Waiting or owing");
```
Comment (justinjmoses) [inline contracts/Collateral.sol:589]: (as above)

File: contracts/Collateral.sol (hunk 21 above)
Comment (jacko125) [reply contracts/Collateral.sol:548]: I believe this was the expectation that fee reclamation on L2 isn't needed or going to be turned on in the future.

File: contracts/Collateral.sol (hunk 22 above)
Comment (barrasso) [reply contracts/Collateral.sol:589]: From jacko: 
> I believe this was the expectation that fee reclamation on L2 isn't needed or going to be turned on in the future.

https://github.com/Synthetixio/synthetix/pull/1448/files#r699881457

File: contracts/Collateral.sol (hunk 20 above)
Comment (barrasso) [reply contracts/Collateral.sol:502]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/93476f10395055003362a5a7bc272b3dd1918028

File: contracts/Collateral.sol (hunk 19 above)
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/93476f10395055003362a5a7bc272b3dd1918028

File: contracts/Collateral.sol (hunk 18 above)
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/93476f10395055003362a5a7bc272b3dd1918028

File: contracts/Collateral.sol (hunk 17 above)
Comment (barrasso) [reply contracts/Collateral.sol]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/93476f10395055003362a5a7bc272b3dd1918028

File: test/integration/utils/loans.js (hunk 16 above)
Comment (barrasso) [reply test/integration/utils/loans.js]: Resolved: https://github.com/Synthetixio/synthetix/pull/1448/commits/6da4a6499cd12d195b157a5ef35443582ce94144

File: contracts/Collateral.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (barrasso) [reply contracts/Collateral.sol:524]: Resolved:

- https://github.com/Synthetixio/synthetix/pull/1448/commits/a1bc6d4af797aa257256accdfd44b23756adc854
- https://github.com/Synthetixio/synthetix/pull/1448/commits/0ffcfc16010d55159ea58d3964853311ad517cac

**Comment by barrasso** [comment]
For `ovm` deployment:

Make sure `interactionDelay` is set to `0`.

Also, ensure the CollateralManager has all Collateral contracts added

File: publish/releases.json [hunk 23]
```diff
@@ -293,6 +293,21 @@
 			"layer": "base",
 			"released": "base"
 		},
+		{
+			"sip": 135,
+			"layer": "both",
```
Comment (justinjmoses) [inline publish/releases.json:298]: this should be `ovm` right - we don't want to schedule a deploy on both layers.

Comment (justinjmoses) [inline publish/releases.json:298]: Instead of this, I suggest targeting `layer: ovm` and for the SystemSettings configs that are loans specific, put them in a conditional that checks if the SystemSettings contract interface has the function you expect (loaded up from the ABI). We don't want to tackle a multilayer deployment just for system setting configuration.

**Comment by barrasso** [comment]
Some gas usage statistics for L2 shorts:

- `short.open`: 9366958

- `short.withdraw`: 4044849

- `short.draw`: 5553515

- `short.close`: 4548746

File: publish/releases.json [hunk 24]
```diff
@@ -293,6 +293,18 @@
...
+			"sip": 135,
+			"layer": "ovm",
+			"sources": [
+				"CollateralManager",
+				"CollateralShort",
+				"CollateralUtil",
+				"MixinSystemSettings",
+				"ShortingRewards",
+				"SystemSettings"
+			]
```
Comment (justinjmoses) [inline publish/releases.json:306]: you will also need to add `SynthsUSD`, `ETH` and `BTC` here as you want them to be instances of `MultiCollateralShort`.

File: contracts/MixinSystemSettings.sol [hunk 25]
```diff
@@ -138,4 +144,52 @@ contract MixinSystemSettings is MixinResolver {
     function getEtherWrapperBurnFeeRate() internal view returns (uint) {
         return flexibleStorage().getUIntValue(SETTING_CONTRACT_NAME, SETTING_ETHER_WRAPPER_BURN_FEE_RATE);
     }
+
+    function getMinCratio(address collateral) internal view returns (uint) {
+        return
+            flexibleStorage().getUIntValue(
+                SETTING_CONTRACT_NAME,
+                keccak256(abi.encodePacked(SETTING_MIN_CRATIO, collateral))
```
Comment (justinjmoses) [inline contracts/MixinSystemSettings.sol:152]: [np] would it be worth using a private function to do the `keccak`? It might reduce the contract size impact of all these.

Comment (justinjmoses) [reply contracts/MixinSystemSettings.sol:152]: nvm - did some testing and seems like it doesn't improve.

**Review by justinjmoses — APPROVED** [review]
:100:
